### ISSUE TITLE
nixpkgs.{system,localSystem,crossSystem} values are useless

### ISSUE THREAD

**Issue by roberth** [NixOS/nixpkgs#49765]
## Issue description

Of course these options serve a purpose inside the `nixpkgs.nix` module, to set `pkgs`. However, these options tend to be used elsewhere, even though they are hard to use, leading to problems with cross compiled NixOS.

For this reason, I suggest that we make these config values unusable by providing an appropriate error message showing what else to do.

### Why aren't they useful

`system` is a legacy option that will now tend to be set to its default, builtins.currentSystem.

`localSystem` will only consistently represent the system where stuff was built. This is typically not useful information, yet it is used erroneously in some modules.

`crossSystem` *is* meaningful, but is not always set.

### What can we do

Maybe `system` can be fixed by inspecting `options.nixpkgs.localSystem` *and* `options.nixpkgs.crossSystem`. However, it is hard to reason about all possible uses of the module system and this feels like laying down a minefield for further changes to the `nixpkgs.nix` module.

A better solution seems to be to just treat these options as write-only. Users will set them and no other module than `nixpkgs.nix`  can use its value. We can then, by means of an error message, direct module authors to the right value. This should probably be `pkgs.targetPlatform`. I don't think NixOS needs to provide its own way to determine its platform. `pkgs` can do that for NixOS just fine.

### How to implement

Simply set all three options to `private`...

The module system does not yet have an [option parameter](https://github.com/NixOS/nixpkgs/blob/e338d801e207a888e3125d88108e8166c94034ac/lib/options.nix#L29) to make it private. A variation of private options can be implemented by adding an extra attribute `internalValue` [here](https://github.com/NixOS/nixpkgs/blob/e338d801e207a888e3125d88108e8166c94034ac/lib/modules.nix#L336) and changing `value` to an error when the option is private. This relies on the implicit agreement that modules don't look at other modules `internalValue`s, which seems reasonable. Compared to *really* enforcing encapsulation, such an implicit agreement is quite useful for debugging, because the value is always around for you to inspect.

#### Overhaul the `nixpkgs` options once more

So we ask people to set
 - nothing at all to use `builtins.currentSystem`
 - `nixpkgs.targetSystem` to set the system of the resulting NixOS
 - `nixpkgs.hostSystem` to set the system to build it on

Boils down to the same thing of course, but now we have a `nixpkgs.targetSystem` option that is actually useful.

I don't know how the old `nixpkgs.system` fits into this.
This means we'll need to reevaluate all the corner cases once more.
It will probably still confuse people and cause them to use `nixpkgs.hostSystem` which will then supposedly fail as it should by being `null` most of the time.

**Comment by coretemp** [comment]
I think the amount of time invested in the development of this feature (cross-compilation) along with the confusion it has created related to the documentation that comes with it is too high myself. 

If you invest a few hundred hours in a feature, documenting it for at least 5 hours seems appropriate to me.

**Comment by roberth** [comment]
Hey coretemp. I too had to look up how to use the cross compilation and I understand your frustration. In particular I find the documentation for the distinction between a system string and { system : string, config : ? } and such lacking.

That said I do appreciate the amount of effort that it takes to work on a such a complicated project. I have been working on some improving some of the glue, like adding `pkgs.nixos`, and I find that the documentation is a significant part of the work. I imagine that the cross compilation is similar in this regard, except much much more involved. Because of its size, I understand that documentation has suffered, but I absolutely agree that it's not done until the documentation is in a good shape.

I am glad to see that you filed an issue, #49510. If there's anything else you think must be improved, please file another issue.

**Comment by Ericson2314** [comment]
@roberth yes the `crossSystem` interface sucks for precisely the reason that there is no easy consistent way to get the build-time and run-time platforms.

@coretemp I did *not* invent such a thing. `crossSystem` already existed for years. I kept it around to not break existing documentation (at the time, the old wiki). I regret doing this; that documentation is little read now anyways.

All that said, I think the solution is to change the arguments to Nixpkgs itself (with some sort of back compat provision) and then change the NixOS interface accordingly.

**Comment by Ericson2314** [comment]
@matthewbauer I recall using `pkgs.anything` could cause infinite recursion problems maybe? That's why a proper fix would definitely be preferred.

**Comment by matthewbauer** [comment]
It shouldn't in NixOS modules unless something very weird is going on. You have to use the final "pkgs" - not config.nipkgs.pkgs though.

I've found a couple usages of localSystem but it seems pretty rare (dysnomia and containers). Were there any others?

**Comment by Ericson2314** [comment]
> You have to use the final "pkgs" - not config.nipkgs.pkgs though.

Oh, ok. Well that would be great to use `pkgs` everywhere and fix remove `system` from https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/eval-config.nix#L11-L12

**Comment by coretemp** [comment]
@Ericson2314 OK, thank you for that clarification.

**Comment by xaverdh** [comment]
We still pass around `system` in `eval-config`, so I think this is not stale..

**Comment by xaverdh** [comment]
just dropping a thought that was floating around in my brain:
`nixpkgs.targetSystem`, which is currently `nixpkgs.crossSystem.system` should probably be set by `nixos-generate-config` in `hardware-configurationx.nix` since its a static property of the target system.

**Comment by alarsyo** [comment]
Still important

**Comment by roberth** [comment]
> just dropping a thought that was floating around in my brain: `nixpkgs.targetSystem`, which is currently `nixpkgs.crossSystem.system` should probably be set by `nixos-generate-config` in `hardware-configurationx.nix` since its a static property of the target system.

Done in
 - https://github.com/NixOS/nixpkgs/pull/177672

while the `system` parameter becomes optional in favor of the well-defined in-module variables in
 - https://github.com/NixOS/nixpkgs/pull/177012

This should become the new way of doing things in NixOS, but I'm taking a very conservative approach so as to validate the solution and to avoid any unnecessary disruption.

Nixpkgs would also benefit from such a change, allowing `buildPlatform` and `hostPlatform` to be set directly, although there the pain isn't felt as much because all consumers already have access to well-defined variables to read from. It's just the nixpkgs invocation that's a little odd.

**Comment by arianvp** [comment]
I'm kind of confused why we started using `hostPlatform/buildPlatform` names in NixOS whilst `nixpkgs` itself still uses `localSystem/crossSystem` as the names.   This really tripped me up.  Which we deprecated `system` but kept the `localSystem,crossSystem` names.

**Comment by roberth** [comment]
Fwiw, `system` was soft-deprecated before I soft-deprecated `localSystem`/`crossSystem` (so I'm not actually 100% sure I understand your comment).

Nonetheless your criticism applies. In retrospect, I should have started with Nixpkgs' parameters first, and make the NixOS change after.
I don't have the capacity to work on this now, but feel free to apply similar changes to the Nixpkgs entrypoints.

**Comment by arianvp** [comment]
Thanks for the context! I  think i'll add a note in the code just to clarify!

**Comment by siraben** [comment]
Bumping this again. I think it's worth discussing since we would like to have uniform terminology. I was surprised by the naming personally.
